### PR TITLE
feat: support exiting early from IterateHierarchy method

### PR DESCRIPTION
**Pull request by alexmt**
Signed-off-by: Alexander Matyushentsev <AMatyushentsev@gmail.com>

### CODE REVIEW

File: pkg/cache/cluster.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ type ClusterCache interface {
 	// FindResources returns resources that matches given list of predicates from specified namespace or everywhere if specified namespace is empty
 	FindResources(namespace string, predicates ...func(r *Resource) bool) map[kube.ResourceKey]*Resource
 	// IterateHierarchy iterates resource tree starting from the specified top level resource and executes callback for each resource in the tree
-	IterateHierarchy(key kube.ResourceKey, action func(resource *Resource, namespaceResources map[kube.ResourceKey]*Resource))
+	IterateHierarchy(key kube.ResourceKey, action func(resource *Resource, namespaceResources map[kube.ResourceKey]*Resource) bool)
```
Comment (jessesuen) [inline pkg/cache/cluster.go:104]: Can you document what the bool return value of the action callback is supposed to indicate?

File: pkg/cache/cluster.go [hunk 2]
```diff
@@ -846,14 +846,15 @@ func (c *clusterCache) IterateHierarchy(key kube.ResourceKey, action func(resour
...
 				child := children[0]
-				action(child, nsNodes)
-				child.iterateChildren(nsNodes, map[kube.ResourceKey]bool{res.ResourceKey(): true}, func(err error, child *Resource, namespaceResources map[kube.ResourceKey]*Resource) {
-					if err != nil {
-						c.log.V(2).Info(err.Error())
-						return
-					}
-					action(child, namespaceResources)
-				})
+				if action(child, nsNodes) {
```
Comment (jessesuen) [inline pkg/cache/cluster.go:849]: `action()` is called on line 831, but I don't see the return value being checked. Is that expected?

Comment (alexmt) [reply pkg/cache/cluster.go:852]: Thank you! This is a bug. Added missing check and unit test

File: pkg/cache/cluster.go (hunk 1 above)
Comment (alexmt) [reply pkg/cache/cluster.go:105]: done

**Comment by alexmt** [comment]
Thank you for review @jessesuen , PTAL

**Review by jessesuen — APPROVED** [review]
LGTM

**Pull request merged**